### ISSUE TITLE
Accomodate multiple components in a route?

### ISSUE THREAD

**Issue by thebritican** [relay-tools/react-router-relay#48]
In the sidebar example, multiple components are specified for a route.
https://github.com/rackt/react-router/blob/master/examples/sidebar/app.js#L102

Should this be possible with relay? Would the `queries` parameter need to accept an object mapping queries to the multiple components? Like:

```jsx
<Route components={{sidebar: Sidebar, content: Content}}
       queries={{sidebar: sidebarQueries, content: contentQueries}}>...</Route>
```

 Or, perhaps the queries could be available for all the (relay) components.
```jsx
<Route components={{sidebar: Sidebar, content: Content}}
       queries={routeQueries}>...</Route>
```

**Comment by taion** [comment]
It should, but the question is "how".

The former proposal is a little odd in that we're still using the prop `queries`, but assigning it a different meaning from the normal case.

The latter proposal is a bit too inflexible, I think.

**Comment by thebritican** [comment]
Perhaps the structure passed into the route should be constructed with queries as a nested key rather than name matching.

```jsx
const containers = {
  sidebar: { 
    component: Sidebar,
    queries: sidebarQueries
  },
  content: {
    component: Content,
    queries: contentQueries
  }
}
```

Then the `Route` can be passed this structure through some combined prop.

```jsx
<Route containers={containers}>...</Route>
```

And if you wanted to pass any non-relay components to the route, still allow the use of `components`?

```jsx
<Route containers={containers} components={components}>...</Route>

**Comment by taion** [comment]
I don't know that this hooks properly into React Router's functionality around async components.

**Comment by thebritican** [comment]
You're right that this would probably require React Router changes, unless the `Route` components are wrapped by a component in this library and the object values of  `containers` become the component value.

```jsx
const containers = { sidebar: { component: Sidebar, queries: sidebarQueries }, ... }
```
to
```jsx
const containerComponents = { sidebar: Sidebar, ... }
```

The higher order component can then merge `containerComponents` and the other prop, `components` and pass it on to `Route`.

**Comment by taion** [comment]
Thinking about this a bit more, maybe your initial proposal of making `queries` a nested object when specifying `components` on the route is just the best way to go about this.

**Comment by thebritican** [comment]
Ok great, I'm working on a PR now

**Comment by taion** [comment]
Oh, cool. Thanks!

**Comment by thebritican** [comment]
Do you have any tests currently that I can run my solution against?

**Comment by taion** [comment]
Not yet sorry. I haven't had the time. I'm also torn between doing everything with a testing stack I'm familiar with and using Jest per https://github.com/relay-tools/react-router-relay/issues/19 and for consistency w/the facebook/relay repo.

**Comment by vladar** [comment]
Need any help? I am quite interested in this feature.

**Comment by thebritican** [comment]
@vladar That would be great! I got quite sidetracked from this feature by the project using this library. I was in the midst of setting up a relay example with a sidebar; I haven't gotten far in modifying the queries parameter since I couldn't test my implementation.

**Comment by taion** [comment]
BTW worth being aware of https://github.com/rackt/react-router/issues/1968.

**Comment by vladar** [comment]
Yeah, and I guess it also depends on solution for async components as well

**Comment by taion** [comment]
Hmm, I'm deferring this a bit more - it's a little tricky because when creating the element for a multi-component route, I actually don't know _which_ element I am creating, which is relevant when trying to link back to the Relay data.

**Comment by taion** [comment]
Blocked on https://github.com/rackt/react-router/pull/2545.

**Comment by taion** [comment]
This is straightforwardly unblocked now with the upstream changes.

**Comment by colllin** [comment]
Hey, what do you think about making a `Relay.createContainer`-like function for composing a Container with its route-based queries?  I'll call this function `relayRoot` below...

Instead of:

    <Route path="campaigns" component={Campaigns} queries={ViewerQueries} />

It could look something like:

    <Route path="campaigns" component={relayRoot(Campaigns, ViewerQueries)} />

Then it becomes more obvious how to handle named components:

    <Route path="campaigns" components={{
        sidebar: relayRoot(Sidebar, sidebarQueries),
        content: relayRoot(Content, contentQueries),
    }}>...</Route>

And how to handle non-relay components:

    <Route path="campaigns" components={{
        sidebar: Sidebar,
        content: relayRoot(Content, contentQueries),
    }}>...</Route>

**Comment by taion** [comment]
I'd rather just have `queries` be an object with keys matching the component keys.

**Comment by colllin** [comment]
Well... I played around with this idea, and got a simple version working.

```
let relayRoot = (Component, queries) => {
    return class extends React.Component {
        render() {
            return (
                <Relay.RootContainer
                    Component={Component}
                    route={{
                        queries,
                        params: this.props.params,
                        name: Component.displayName || Component.name,
                    }}
                    renderFetched={(data) => (<Component {...this.props} {...data} />)}
                />
            );
            // renderLoading={ /* TODO */ }
            // renderFailure={ /* TODO */ }
        }
    }
};
```

It seems to work with named components.  I haven't tested with non-relay components.  

Using this `relayRoot` function [as described above](https://github.com/relay-tools/react-router-relay/issues/48#issuecomment-170381984), I was actually able to remove `'react-router-relay'` in favor of plain old `'react-router'` and `<Router>`.

@taion Any interest in this idea?

**Comment by taion** [comment]
If it were that easy, we wouldn't have made this package in the first place.

Take a look at @cpojer's writeup on the considerations involved: https://medium.com/@cpojer/relay-and-routing-36b5439bad9.

**Comment by colllin** [comment]
Thanks for the link.  That would have saved me a few hours today.  I realize your package must be doing something more advanced – I just didn't know what.  For anyone else interested, I think that article describes [a solution similar to mine] here:

> This solution will work for most smaller applications but it has one flaw: on every level of the nested route tree, when it finds a Relay component, it will render a new RelayRootContainer. This means that on every step of the route tree, Relay has to fetch all the data for that level until it can proceed to fetch data for the next level in the route tree. Relay is trying to solve the multiple-API-request problem by using GraphQL but if you are using the above solution you will not benefit from Relay’s optimizations.

I'm looking forward to your solution for named components.

**Comment by stevewillard** [comment]
Any update on this? Interested in using this feature of react-router with relay

**Comment by taion** [comment]
I'd be happy to accept a PR. This is not a particularly difficult change to make; it's just not a priority for me at the moment.
